extensions/sftp/client/SFTPClient.h

```diff
@@ -0,0 +1,183 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __SFTP_CLIENT_H__
+#define __SFTP_CLIENT_H__
+
+#include <curl/curl.h>
+#include <libssh2.h>
+#include <libssh2_sftp.h>
+#include <vector>
+#include <iostream>
+#include <string>
+#include <uuid/uuid.h>
+#include <vector>
+
+#include "utils/HTTPClient.h"
+#include "core/logging/Logger.h"
+#include "core/logging/LoggerConfiguration.h"
+#include "properties/Configure.h"
+#include "io/validation.h"
+#include "io/BaseStream.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+
+/**
+ * Initializes and cleans up curl and libssh2 once.
+ * Cleanup will only occur at the end of our execution since we are relying on a static variable.
+ */
+class SFTPClientInitializer {
+ public:
+  static SFTPClientInitializer *getInstance() {
+    static SFTPClientInitializer initializer;
+    return &initializer;
+  }
+  void initialize() {
+
+  }
+ private:
+  SFTPClientInitializer() {
+    libssh2_init(0);
+    curl_global_init(CURL_GLOBAL_DEFAULT);
```

Comment:
curl_global_init is not thread safe. The singleton pattern appears to model that which is in http-curl, which is lazily initialized.

Reply:
Neither is libssh2_init thread-safe.
You're right, I have modeled the SFTPClientInitializer on HTTPClientInitializer from http-curl, and I don't see how it would be more or less thread-safe in itself than that (a block-scope static object provides synchronization). It is even initialized in the same way (lazily, both by the respective client and in the constructor of the Processor that uses that client).
However, now I do realize, that globally it became less thread-safe, because if both initializers would be constructed at the same time on different threads, there would be no synchronization between them, and curl_global_init could indeed be called in an unsafe manner.
If this could happen, we need to create global initializers for these third-party libraries. Creating a separate CurlInitializer and LibSSH2Initializer won't solve the problem, however, because even calling curl_global_init and libssh2_init at the same time is unsafe, because they both potentially call OpenSSL's initialization, which is not thread-safe in itself.
To solve this we need to find a way to synchronize all thirdparty global initializations and deinitializations.
What do you think, @phrocker ?

Reply:
@bakaid Thanks for pointing that out. One option is create a varying dependency tree of lazy initializers -- but I think as you point out that will be error prone.

Extensions were intended to be groups of functionality and one other issue is that the bootstrap script will eventually run out of characters. I'd like to get to a point where more of these extensions that rely on the same dependencies are combined. 

Another option here could be combining sftp and http-curl into a curl extension, and using the HTTPCurlLoader functionality to eagerly initialize these components. We can still have SFTP be an included component within that extension, but then all curl based functionality will be included within a single extension with the ability to pick and choose sub-features as needed. Does that make sense? 

There will be more features that use curl, so having http-curl and other curl reliant extensions doesn't make a ton of sense to me. We can still have the same ability to limit features included and also have the byproduct that initialization can be done via the ObjectFactory ( a component of the class loader that is loaded before all processors start ). 

While we certainly could have the case where something "like this" happens again, I'm not sure we can avoid that as that's a consistent issue amongst Java Classloaders, too. What are your thoughts on this? It can be something we do after this is merged, labeling it a blocker to get it done, but I'll leave that up to you since there is a ton of good work here and I don't want to slow this up for such a change unless you are open to do that. 

I'll give it more thought, but those are two options I can think of that would solve this and other issues that will arise by having the split we have.

Reply:
@phrocker On grouping the extensions by dependencies: from our standpoint that could make sense (it could solve our current problem, for example), as long as extensions do not depend on different sets of multiple external dependencies. The sftp extension depends on curl and libssh2, so we could group it under extensions depending on curl, with an extra attribute of depending on libssh2.
But what would happen if in the future we want to create an extension which depends on libssh2 but does not depend on curl? (I don’t think this particular case will happen, but I’m trying to figure out a general architecture.) In that case we will still have to group it under curl-dependent extensions and make it dependent on curl, as we could not create a different initializer for it that calls libssh2_init, because that could potentially be called in parallel with HTTPCurlLoader.
Even worse, If our extension would depend on libssh2 and libfoobar, we would have to group all extensions depending on libfoobar with the extensions depending both on curl and libssh2, as we couldn’t allow having two initializers calling libfoobar_init.
From the standpoint of the user I would imagine (but I could be terribly wrong in this case), that they do not care about the thirdparty dependencies we use to implement processors - they care about the particular processor and the functionality it brings. They don’t want to have all functionality that could be implemented with curl - they want SFTP, FTP or HTTP.
So to recap: grouping similar processors into the same extensions, especially if they have the same dependencies could make sense on a case-to-case basis but I don’t think that grouping all processors depending on a particular external dependency into the same extension solves our problems or brings much to the table (we could solve running out of letters in bootstrap.sh by other means), but brings its own set of problems.

A way to solve this based on your recommendation perhaps would be to eagerly initialize the components in the loader AND call these initialization functions under a global lock (move the lock in ClassLoader::registerResource up a few lines). This way we wouldn’t have to change the architecture drastically, but still wouldn’t have to group all processors depending on curl (and all other transitive dependencies) in the same extension.

I too will think about this more.